### PR TITLE
Allow to extend how caching JDBC client is created

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    @Inject
+    public DefaultCachingJdbcClientFactory(JdbcClient jdbcClient)
+    {
+        this.jdbcClient = requireNonNull(jdbcClient, "jdbcClient is null");
+    }
+
+    @Override
+    public JdbcClient create()
+    {
+        return new CachingJdbcClient(jdbcClient, new Duration(1, TimeUnit.DAYS), true);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:38]: Where does 1d come from?
sounds like "almost infinity"

use io.prestosql.plugin.jdbc.BaseJdbcConfig#getMetadataCacheTtl here

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+public class DefaultCachingJdbcClientFactory
+        implements CachingJdbcClientFactory
+{
+    private final JdbcClient jdbcClient;
+
+    @Inject
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:29]: Does it mean `CachingJdbcClient` is no longer `@Inject`-ed and https://github.com/prestosql/presto/blob/142c8e20fb14535064667f8a7978525aa6e1f6f5/presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java#L62-L66 ctor should be removed?

or (likely), this is only for per-trnasaction cache?
then what about other cache?

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.jdbc;
+
+public interface CachingJdbcClientFactory
+{
+    JdbcClient create();
```
Comment (kokosing) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java:18]: Should `JdbcClient` become closeable?

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java (hunk 1 above)
Comment (sopel39) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:38]: > sounds like "almost infinity"

That makes sense within single transaction. The value was moved from previous code version (see `JdbcMetadataFactory`)

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java (hunk 2 above)
Comment (sopel39) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:29]: `CachingJdbcClient` is still created by Guava. There are two caches:
1. per-transaction cache
2. global cache (created by Guava) that is cross transactions.

This is similar mechanism to what we have in Hive

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java (hunk 1 above)
Comment (findepi) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:38]: 1d is neither "reasonably short for something" not "infinite"
I am OK not changing this, but maybe "TODO use infinity for in-transaction cache"

Comment (sopel39) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:38]: > I am OK not changing this, but maybe "TODO use infinity for in-transaction cache"

I don't think planner will ask for stats after one day ;)

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.jdbc;
+
+public interface CachingJdbcClientFactory
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java:16]: Guice wants to know what you bind.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java (hunk 3 above)
Comment (findepi) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java:18]: is in-transaction state explicitly closed?

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+import io.airlift.units.Duration;
+
+import javax.inject.Inject;
+
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+public class DefaultCachingJdbcClientFactory
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:24]: "Default...." indicates this is the default impl
"Transactional..." indicates the role -- it's suitable within interface name (which is used for managing Guice binds)

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java (hunk 1 above)
Comment (findepi) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DefaultCachingJdbcClientFactory.java:38]: this is about any caching, not related to stats

if i have a very long running transaction, it does not make sense to flush state after 24h.
if we want to limit trx length, we should do this explciitly (& somewhere else)

**Review by findepi — APPROVED** [review]

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java (hunk 3 above)
Comment (sopel39) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClientFactory.java:18]: > Should JdbcClient become closeable?

You shouldn't close delegate JdbcClient, which is shared

**Pull request merged**